File: python/tests/test_risk.py
"""Unit tests for algoforge.risk (PositionSizer, DrawdownGuard, SignalProcessor).

Run with:
    python -m pytest python/tests/test_risk.py -v
or
    python -m unittest python.tests.test_risk
"""
import unittest
from unittest.mock import MagicMock

from algoforge.risk import (
    DrawdownGuard,
    GuardState,
    PositionSizer,
    SignalProcessor,
    SizingMethod,
)
from algoforge.broker import AccountInfo, OrderType, SymbolInfo, Tick, Order
from algoforge.algorithm import AlgoDecision, AlgoSignal
from algoforge.types import Direction


# ---------------------------------------------------------------------------
# PositionSizer tests
# ---------------------------------------------------------------------------

class TestKellyFraction(unittest.TestCase):

    def test_kelly_basic(self):
        """WR=0.55, RR=2.0 → half-Kelly=0.1625, capped at max_risk=0.02."""
        # edge = 0.55*2 - 0.45 = 0.65; kelly = 0.65/2 = 0.325; half = 0.1625
        result = PositionSizer.kelly_fraction(win_rate=0.55, rr=2.0, max_risk=0.02)
        self.assertAlmostEqual(result, 0.02)   # capped by max_risk

    def test_kelly_uncapped(self):
        """With generous cap the raw half-Kelly is returned."""
        result = PositionSizer.kelly_fraction(win_rate=0.55, rr=2.0, max_risk=0.5)
        self.assertAlmostEqual(result, 0.1625, places=6)

    def test_kelly_floor(self):
        """Negative edge → clamped to minimum floor of 0.005."""
        result = PositionSizer.kelly_fraction(win_rate=0.3, rr=1.0, max_risk=0.02)
        self.assertAlmostEqual(result, 0.005)

    def test_kelly_zero_rr(self):
        """rr=0 → kelly=0, returned as floor 0.005."""
        result = PositionSizer.kelly_fraction(win_rate=0.6, rr=0.0, max_risk=0.02)
        self.assertAlmostEqual(result, 0.005)


class TestNormalizeLots(unittest.TestCase):

    def test_round_to_step(self):
        result = PositionSizer.normalize_lots(1.567, step=0.01, min_lots=0.01, max_lots=100.0)
        self.assertAlmostEqual(result, 1.57)

    def test_clamp_to_min(self):
        result = PositionSizer.normalize_lots(0.001, step=0.01, min_lots=0.01, max_lots=100.0)
        self.assertAlmostEqual(result, 0.01)

    def test_clamp_to_max(self):
        result = PositionSizer.normalize_lots(500.0, step=0.01, min_lots=0.01, max_lots=100.0)
        self.assertAlmostEqual(result, 100.0)

    def test_zero_step_defaults_to_001(self):
        result = PositionSizer.normalize_lots(1.567, step=0.0, min_lots=0.01, max_lots=100.0)
        self.assertAlmostEqual(result, 1.57)

    def test_already_on_step(self):
        result = PositionSizer.normalize_lots(0.10, step=0.01, min_lots=0.01, max_lots=100.0)
        self.assertAlmostEqual(result, 0.10)


class TestSizePosition(unittest.TestCase):

    def _default_sym(self) -> SymbolInfo:
        return SymbolInfo(
            name="EURUSD",
            digits=5,
            point=0.00001,
            contract_size=100_000.0,
            volume_min=0.01,
            volume_max=100.0,
            volume_step=0.01,
        )

    def test_fixed_risk_basic(self):
        """Simple sanity-check: 1% risk on 10k, 10-pip stop."""
        sym = self._default_sym()
        # stop_loss > 0 so stop_dist = abs(1.1100 - 1.1090) = 0.0010
        # stop_in_pts = 0.001 / 0.00001 = 100 pts
        # val_per_point = 100_000 * 0.00001 = 1.0
        # risk_amount = 10_000 * 0.01 = 100
        # raw_lots = 100 / (100 * 1.0) = 1.0
        result = PositionSizer.size_position(
            method=SizingMethod.FIXED_RISK,
            balance=10_000.0,
            entry_price=1.1100,
            stop_loss=1.1090,
            sym=sym,
            atr=0.0010,
            risk_pct=0.01,
            max_lots=100.0,
            min_lots=0.01,
        )
        self.assertAlmostEqual(result.lots, 1.0, places=2)
        self.assertFalse(result.capped)
        self.assertAlmostEqual(result.stop_distance, 0.0010, places=6)

    def test_kelly_sizing(self):
        """Kelly method should produce a smaller lot than fixed at same risk_pct."""
        sym = self._default_sym()
        fixed = PositionSizer.size_position(
            method=SizingMethod.FIXED_RISK,
            balance=10_000.0,
            entry_price=1.1100,
            stop_loss=1.1090,
            sym=sym,
            atr=0.0010,
            risk_pct=0.02,
            max_lots=100.0,
            min_lots=0.01,
        )
        kelly = PositionSizer.size_position(
            method=SizingMethod.KELLY,
            balance=10_000.0,
            entry_price=1.1100,
            stop_loss=1.1090,
            sym=sym,
            atr=0.0010,
            risk_pct=0.02,
            max_lots=100.0,
            min_lots=0.01,
            win_rate=0.55,
            rr_ratio=2.0,
        )
        # Kelly caps at max_risk=0.02 (same as fixed) → equal in this case
        # but kelly half is 0.1625 → still capped to 0.02 → same lots
        self.assertGreater(kelly.lots, 0.0)

    def test_atr_stop_used_when_no_sl(self):
        """When stop_loss=0 the ATR-derived stop should be used."""
        sym = self._default_sym()
        result = PositionSizer.size_position(
            method=SizingMethod.FIXED_RISK,
            balance=10_000.0,
            entry_price=1.1100,
            stop_loss=0.0,       # no explicit SL
            sym=sym,
            atr=0.0020,
            risk_pct=0.01,
            max_lots=100.0,
            min_lots=0.01,
        )
        expected_stop = 0.0020 * 1.5
        self.assertAlmostEqual(result.stop_distance, expected_stop, places=6)

    def test_capped_flag(self):
        """Ridiculously small stop → raw lots explodes → capped=True."""
        sym = self._default_sym()
        result = PositionSizer.size_position(
            method=SizingMethod.FIXED_RISK,
            balance=100_000.0,
            entry_price=1.1100,
            stop_loss=1.10999,   # 0.1 pip stop
            sym=sym,
            atr=0.00001,
            risk_pct=0.05,
            max_lots=100.0,
            min_lots=0.01,
        )
        self.assertTrue(result.capped)
        self.assertEqual(result.lots, 100.0)


# ---------------------------------------------------------------------------
# DrawdownGuard tests
# ---------------------------------------------------------------------------

class TestDrawdownGuard(unittest.TestCase):

    def test_green_on_init(self):
        guard = DrawdownGuard()
        guard.initialise(balance=10_000, equity=10_000)
        self.assertEqual(guard.state, GuardState.GREEN)
        self.assertTrue(guard.trading_allowed)
        self.assertEqual(guard.halt_count, 0)

    def test_red_on_daily_breach(self):
        """Equity drops to 9 400 → daily loss = −6 % > 5 % limit → RED."""
        guard = DrawdownGuard(daily_limit=0.05, total_limit=0.15)
        guard.initialise(balance=10_000, equity=10_000)
        status = guard.check(equity=9_400, balance=10_000)
        self.assertEqual(status.state, GuardState.RED)
        self.assertFalse(status.trading_allowed)
        self.assertEqual(guard.halt_count, 1)
        self.assertIn("Daily", status.reason)

    def test_yellow_on_warning_threshold(self):
        """Equity at 9 600 → daily loss = −4 % which is ≥ 75 %×5 % = 3.75 % → YELLOW."""
        guard = DrawdownGuard(daily_limit=0.05, total_limit=0.15)
        guard.initialise(balance=10_000, equity=10_000)
        status = guard.check(equity=9_600, balance=10_000)
        self.assertEqual(status.state, GuardState.YELLOW)
        self.assertTrue(status.trading_allowed)   # YELLOW still allows trading
        self.assertEqual(guard.halt_count, 0)

    def test_still_green_below_warn_threshold(self):
        """Equity at 9 650 → daily loss = −3.5 % which is < 3.75 % → stays GREEN."""
        guard = DrawdownGuard(daily_limit=0.05, total_limit=0.15)
        guard.initialise(balance=10_000, equity=10_000)
        status = guard.check(equity=9_650, balance=10_000)
        self.assertEqual(status.state, GuardState.GREEN)

    def test_red_on_total_drawdown(self):
        """Peak rises during the session, then >15 % total drawdown → RED."""
        guard = DrawdownGuard(daily_limit=0.05, total_limit=0.15)
        guard.initialise(balance=10_000, equity=10_000)
        # Simulate equity rising to 12_000 (sets new peak)
        guard.check(equity=12_000, balance=10_000)
        self.assertAlmostEqual(guard._peak, 12_000)
        # Now equity retreats — total dd = (12_000 - 10_100) / 12_000 ≈ 0.158 > 0.15
        # daily P&L = 10_100 - 10_000 = +100 (+1%) so daily limit is NOT triggered
        status = guard.check(equity=10_100, balance=10_000)
        self.assertEqual(status.state, GuardState.RED)
        self.assertFalse(status.trading_allowed)

    def test_peak_updates(self):
        """New equity high should update peak."""
        guard = DrawdownGuard()
        guard.initialise(balance=10_000, equity=10_000)
        guard.check(equity=11_000, balance=10_000)   # new high
        self.assertAlmostEqual(guard._peak, 11_000)

    def test_no_double_halt_count(self):
        """Calling check repeatedly while RED should not keep incrementing halt_count."""
        guard = DrawdownGuard(daily_limit=0.05)
        guard.initialise(balance=10_000, equity=10_000)
        guard.check(equity=9_400, balance=10_000)
        guard.check(equity=9_300, balance=10_000)
        guard.check(equity=9_200, balance=10_000)
        self.assertEqual(guard.halt_count, 1)

    def test_metrics_values(self):
        """Check daily_pnl and total_dd are calculated correctly."""
        guard = DrawdownGuard()
        guard.initialise(balance=10_000, equity=10_000)
        status = guard.check(equity=9_800, balance=10_000)
        self.assertAlmostEqual(status.daily_pnl, -200.0, places=1)
        self.assertAlmostEqual(status.daily_pnl_pct, -0.02, places=4)
        self.assertAlmostEqual(status.total_dd_pct, 0.02, places=4)


# ---------------------------------------------------------------------------
# SignalProcessor tests (using a mock broker)
# ---------------------------------------------------------------------------

def _make_broker(balance=10_000.0, atr=0.0020) -> MagicMock:
    broker = MagicMock()
    broker.get_account.return_value = AccountInfo(
        balance=balance, equity=balance
    )
    broker.get_tick.return_value = Tick(
        symbol="EURUSD", bid=1.10990, ask=1.11000
    )
    broker.get_symbol_info.return_value = SymbolInfo(
        name="EURUSD",
        point=0.00001,
        contract_size=100_000.0,
        volume_min=0.01,
        volume_max=100.0,
        volume_step=0.01,
    )
    # Order with ticket > 0 signals success
    broker.place_order.return_value = Order(ticket=12345, symbol="EURUSD")
    return broker


class TestSignalProcessor(unittest.TestCase):

    def _make_buy_decision(self, confidence=0.70, atr=0.0020) -> AlgoDecision:
        return AlgoDecision(
            signal=AlgoSignal.BUY,
            symbol="EURUSD",
            direction=Direction.LONG,
            confidence=confidence,
            reason="test buy",
            atr=atr,
        )

    def _make_sell_decision(self, confidence=0.70, atr=0.0020) -> AlgoDecision:
        return AlgoDecision(
            signal=AlgoSignal.SELL,
            symbol="EURUSD",
            direction=Direction.SHORT,
            confidence=confidence,
            reason="test sell",
            atr=atr,
        )

    def test_process_buy_success(self):
        broker = _make_broker()
        sp = SignalProcessor(broker, max_risk=0.01, cooldown_sec=0)
        result = sp.process(self._make_buy_decision())
        self.assertTrue(result)
        self.assertEqual(sp.processed, 1)
        self.assertEqual(sp.rejected, 0)
        # Verify a BUY order was placed
        broker.place_order.assert_called_once()
        args = broker.place_order.call_args[0]
        self.assertEqual(args[1], OrderType.BUY)

    def test_process_sell_success(self):
        broker = _make_broker()
        sp = SignalProcessor(broker, max_risk=0.01, cooldown_sec=0)
        result = sp.process(self._make_sell_decision())
        self.assertTrue(result)
        args = broker.place_order.call_args[0]
        self.assertEqual(args[1], OrderType.SELL)

    def test_non_actionable_rejected(self):
        """Low-confidence decision should be rejected without broker calls."""
        broker = _make_broker()
        sp = SignalProcessor(broker, max_risk=0.01, cooldown_sec=0)
        decision = self._make_buy_decision(confidence=0.40)   # below 0.55
        result = sp.process(decision)
        self.assertFalse(result)
        self.assertEqual(sp.rejected, 1)
        broker.place_order.assert_not_called()

    def test_none_signal_rejected(self):
        broker = _make_broker()
        sp = SignalProcessor(broker, max_risk=0.01, cooldown_sec=0)
        result = sp.process(AlgoDecision.none("EURUSD"))
        self.assertFalse(result)
        broker.place_order.assert_not_called()

    def test_cooldown_blocks_second_signal(self):
        """Second signal within cooldown window should be rejected."""
        broker = _make_broker()
        sp = SignalProcessor(broker, max_risk=0.01, cooldown_sec=300)
        sp.process(self._make_buy_decision())   # first → accepted
        result = sp.process(self._make_buy_decision())  # second → cooldown
        self.assertFalse(result)
        self.assertEqual(sp.processed, 1)
        self.assertEqual(sp.rejected, 1)

    def test_opposite_direction_not_blocked_by_cooldown(self):
        """A SELL after a BUY cooldown should use a different key → not blocked."""
        broker = _make_broker()
        # Reset the order mock for second call too
        broker.place_order.side_effect = [
            Order(ticket=1, symbol="EURUSD"),
            Order(ticket=2, symbol="EURUSD"),
        ]
        sp = SignalProcessor(broker, max_risk=0.01, cooldown_sec=300)
        r1 = sp.process(self._make_buy_decision())
        r2 = sp.process(self._make_sell_decision())   # different key → allowed
        self.assertTrue(r1)
        self.assertTrue(r2)
        self.assertEqual(sp.processed, 2)

    def test_failed_order_counted_as_rejected(self):
        """If broker returns ticket=0, process() returns False and increments rejected."""
        broker = _make_broker()
        broker.place_order.return_value = Order(ticket=0, symbol="EURUSD")
        sp = SignalProcessor(broker, max_risk=0.01, cooldown_sec=0)
        result = sp.process(self._make_buy_decision())
        self.assertFalse(result)
        self.assertEqual(sp.processed, 0)
        self.assertEqual(sp.rejected, 1)

    def test_sl_tp_placement_buy(self):
        """For a BUY: sl = ask - atr*1.5, tp = ask + atr*3.0."""
        broker = _make_broker()
        sp = SignalProcessor(broker, max_risk=0.01, cooldown_sec=0)
        atr = 0.0020
        sp.process(self._make_buy_decision(atr=atr))
        _, kwargs = broker.place_order.call_args
        args = broker.place_order.call_args[0]
        # args: symbol, order_type, lots, price, sl, tp, magic, comment
        ask = 1.11000
        expected_sl = ask - atr * 1.5
        expected_tp = ask + atr * 3.0
        self.assertAlmostEqual(args[4], expected_sl, places=5)
        self.assertAlmostEqual(args[5], expected_tp, places=5)

    def test_sl_tp_placement_sell(self):
        """For a SELL: sl = bid + atr*1.5, tp = bid - atr*3.0."""
        broker = _make_broker()
        sp = SignalProcessor(broker, max_risk=0.01, cooldown_sec=0)
        atr = 0.0020
        sp.process(self._make_sell_decision(atr=atr))
        args = broker.place_order.call_args[0]
        bid = 1.10990
        expected_sl = bid + atr * 1.5
        expected_tp = bid - atr * 3.0
        self.assertAlmostEqual(args[4], expected_sl, places=5)
        self.assertAlmostEqual(args[5], expected_tp, places=5)

    def test_atr_fallback_when_zero(self):
        """If decision.atr==0, atr defaults to ask*0.001."""
        broker = _make_broker()
        sp = SignalProcessor(broker, max_risk=0.01, cooldown_sec=0)
        decision = AlgoDecision(
            signal=AlgoSignal.BUY,
            symbol="EURUSD",
            direction=Direction.LONG,
            confidence=0.70,
            reason="no-atr",
            atr=0.0,   # zero
        )
        sp.process(decision)
        args = broker.place_order.call_args[0]
        ask     = 1.11000
        atr_fb  = ask * 0.001          # fallback
        expected_sl = ask - atr_fb * 1.5
        self.assertAlmostEqual(args[4], expected_sl, places=5)


if __name__ == "__main__":
    unittest.main()

File: c/tests/test_risk.c
/**
 * AlgoForge — c/tests/test_risk.c
 * Round 10: Unit tests for risk / position-sizing / drawdown guard.
 *
 * Compile (from repo root):
 *   cc -std=c11 -Wall -Wextra -lm \
 *      c/tests/test_risk.c -o test_risk && ./test_risk
 *
 * Signal processor is NOT tested here because it requires a live broker.
 * Drawdown-guard and position-sizer are fully standalone.
 */

#include <stdio.h>
#include <math.h>
#include <string.h>

#include "../include/af_risk.h"
#include "../src/risk.c"
#include "../src/drawdown_guard.c"

/* ──────────────────────────────────────────────────────────────────────────
 * Helpers
 * ────────────────────────────────────────────────────────────────────────── */
static int g_tests  = 0;
static int g_passed = 0;
static int g_failed = 0;

#define ASSERT_NEAR(label, got, expected, tol)                              \
    do {                                                                     \
        g_tests++;                                                           \
        double _g = (got), _e = (expected), _t = (tol);                    \
        if (fabs(_g - _e) <= _t) {                                          \
            g_passed++;                                                      \
            printf("  PASS  %-42s  got=%.6f\n", (label), _g);              \
        } else {                                                             \
            g_failed++;                                                      \
            printf("  FAIL  %-42s  got=%.6f  expected≈%.6f (tol=%.6f)\n", \
                   (label), _g, _e, _t);                                    \
        }                                                                    \
    } while (0)

#define ASSERT_INT(label, got, expected)                                     \
    do {                                                                     \
        g_tests++;                                                           \
        int _g = (got), _e = (expected);                                    \
        if (_g == _e) {                                                      \
            g_passed++;                                                      \
            printf("  PASS  %-42s  got=%d\n", (label), _g);                \
        } else {                                                             \
            g_failed++;                                                      \
            printf("  FAIL  %-42s  got=%d  expected=%d\n",                 \
                   (label), _g, _e);                                         \
        }                                                                    \
    } while (0)

/* ──────────────────────────────────────────────────────────────────────────
 * Test 1 — af_kelly_fraction
 * ────────────────────────────────────────────────────────────────────────── */
static void test_kelly(void)
{
    printf("\n=== af_kelly_fraction ===\n");

    /*
     * win_rate=0.55, rr=2.0
     * edge = 0.55*2.0 - 0.45 = 1.10 - 0.45 = 0.65
     * kelly = 0.65 / 2.0 = 0.325   (full Kelly)
     * half  = 0.1625
     * max_risk=0.10  → capped to 0.10
     */
    double k1 = af_kelly_fraction(0.55, 2.0, 0.10);
    ASSERT_NEAR("WR=0.55 RR=2.0 cap=0.10 → 0.10", k1, 0.10, 1e-9);

    /* Uncapped: max_risk=0.20 → half-Kelly = 0.1625 */
    double k2 = af_kelly_fraction(0.55, 2.0, 0.20);
    ASSERT_NEAR("WR=0.55 RR=2.0 cap=0.20 → 0.1625", k2, 0.1625, 1e-9);

    /* Negative edge: WR=0.40, RR=1.0 → edge=-0.20 → kelly=-0.20 → half=-0.10
     * floored to 0.005 */
    double k3 = af_kelly_fraction(0.40, 1.0, 0.10);
    ASSERT_NEAR("WR=0.40 RR=1.0 (neg edge) → floor 0.005", k3, 0.005, 1e-9);

    /* RR=0 guard: result should be 0.005 (floored) */
    double k4 = af_kelly_fraction(0.60, 0.0, 0.05);
    ASSERT_NEAR("RR=0 guard → 0.005", k4, 0.005, 1e-9);
}

/* ──────────────────────────────────────────────────────────────────────────
 * Test 2 — af_normalize_lots
 * ────────────────────────────────────────────────────────────────────────── */
static void test_normalize(void)
{
    printf("\n=== af_normalize_lots ===\n");

    /* Basic rounding: 0.123 with step 0.01 → 0.12 */
    ASSERT_NEAR("0.123 step=0.01 → 0.12",
                af_normalize_lots(0.123, 0.01, 0.01, 100.0), 0.12, 1e-9);

    /* Rounding up: 0.126 with step 0.01 → 0.13 */
    ASSERT_NEAR("0.126 step=0.01 → 0.13",
                af_normalize_lots(0.126, 0.01, 0.01, 100.0), 0.13, 1e-9);

    /* Clamped to min: 0.001 with min=0.01 → 0.01 */
    ASSERT_NEAR("0.001 clamped to min=0.01",
                af_normalize_lots(0.001, 0.01, 0.01, 100.0), 0.01, 1e-9);

    /* Clamped to max: 200 with max=100 → 100 */
    ASSERT_NEAR("200.0 clamped to max=100.0",
                af_normalize_lots(200.0, 0.01, 0.01, 100.0), 100.0, 1e-9);

    /* Step=0 defaults to 0.01: 0.555 → 0.56 */
    ASSERT_NEAR("step=0 defaults to 0.01 → 0.56",
                af_normalize_lots(0.555, 0.0, 0.01, 100.0), 0.56, 1e-9);

    /* Step=0.1: 0.74 → 0.7 */
    ASSERT_NEAR("0.74 step=0.1 → 0.70",
                af_normalize_lots(0.74, 0.1, 0.01, 100.0), 0.7, 1e-9);
}

/* ──────────────────────────────────────────────────────────────────────────
 * Test 3 — af_size_position
 * ────────────────────────────────────────────────────────────────────────── */
static void test_size_position(void)
{
    printf("\n=== af_size_position ===\n");

    af_symbol_info_t sym;
    memset(&sym, 0, sizeof(sym));
    snprintf(sym.name, sizeof(sym.name), "EURUSD");
    sym.digits        = 5;
    sym.point         = 0.00001;
    sym.contract_size = 100000.0;
    sym.volume_min    = 0.01;
    sym.volume_max    = 100.0;
    sym.volume_step   = 0.01;

    af_size_result_t res;
    af_error_t rc;

    /* ── Case 1: FIXED_RISK, 1 % of $10 000, entry=1.10000, SL=1.09000
     * stop_dist = |1.10000 - 1.09000| = 0.01000
     * stop_in_pts = 0.01 / 0.00001 = 1000
     * val_per_point = 100000 * 0.00001 = 1.0
     * risk_amount = 10000 * 0.01 = 100
     * raw_lots = 100 / (1000 * 1.0) = 0.10
     */
    rc = af_size_position(AF_SIZE_FIXED_RISK,
                          10000.0, 1.10000, 1.09000,
                          &sym, 0.0, 0.01,
                          100.0, 0.01,
                          0.5, 2.0, &res);
    ASSERT_INT("FIXED_RISK rc=AF_OK", rc, AF_OK);
    ASSERT_NEAR("FIXED_RISK lots=0.10",        res.lots,          0.10, 0.005);
    ASSERT_NEAR("FIXED_RISK risk_amount≈100",  res.risk_amount,  100.0, 1.0);
    ASSERT_NEAR("FIXED_RISK risk_pct≈0.01",    res.risk_pct,     0.01,  0.001);
    ASSERT_NEAR("FIXED_RISK stop_dist=0.01",   res.stop_distance, 0.01, 1e-9);

    /* ── Case 2: ATR-based stop (stop_loss=0, atr provided)
     * stop_dist = 0.00200 * 1.5 = 0.00300
     * stop_in_pts = 300
     * risk_amount = 10000 * 0.01 = 100
     * raw_lots = 100 / (300 * 1.0) = 0.3333
     * normalised → 0.33
     */
    rc = af_size_position(AF_SIZE_ATR_RISK,
                          10000.0, 1.10000, 0.0,
                          &sym, 0.00200, 0.01,
                          100.0, 0.01,
                          0.5, 2.0, &res);
    ASSERT_INT("ATR_RISK rc=AF_OK", rc, AF_OK);
    ASSERT_NEAR("ATR_RISK stop_dist=0.003",    res.stop_distance, 0.003,  1e-9);
    ASSERT_NEAR("ATR_RISK lots≈0.33",          res.lots,          0.33,   0.02);

    /* ── Case 3: KELLY method
     * WR=0.55, RR=2.0 → half-Kelly=0.1625, cap=0.02 → effective risk=0.02
     * risk_amount = 200; same stop_dist=0.01 → lots=0.20
     */
    rc = af_size_position(AF_SIZE_KELLY,
                          10000.0, 1.10000, 1.09000,
                          &sym, 0.0, 0.02,   /* max_risk=0.02 */
                          100.0, 0.01,
                          0.55, 2.0, &res);
    ASSERT_INT("KELLY rc=AF_OK", rc, AF_OK);
    ASSERT_NEAR("KELLY lots=0.20", res.lots, 0.20, 0.01);

    /* ── Case 4: invalid param (NULL sym) */
    rc = af_size_position(AF_SIZE_FIXED_RISK,
                          10000.0, 1.10000, 1.09000,
                          NULL, 0.0, 0.01,
                          100.0, 0.01,
                          0.5, 2.0, &res);
    ASSERT_INT("NULL sym → AF_ERR_INVALID_PARAM", rc, AF_ERR_INVALID_PARAM);

    /* ── Case 5: negative balance */
    rc = af_size_position(AF_SIZE_FIXED_RISK,
                          -1000.0, 1.10000, 1.09000,
                          &sym, 0.0, 0.01,
                          100.0, 0.01,
                          0.5, 2.0, &res);
    ASSERT_INT("balance<0 → AF_ERR_INVALID_PARAM", rc, AF_ERR_INVALID_PARAM);
}

/* ──────────────────────────────────────────────────────────────────────────
 * Test 4 — af_drawdown_guard_check
 * ────────────────────────────────────────────────────────────────────────── */
static void test_drawdown_guard(void)
{
    printf("\n=== af_drawdown_guard_check ===\n");

    af_drawdown_guard_t g;
    af_guard_status_t   s;

    /* Init: balance=10000, equity=10000 */
    af_drawdown_guard_init(&g, 10000.0, 10000.0,
                           AF_DRAWDOWN_DAILY_LIMIT,
                           AF_DRAWDOWN_TOTAL_LIMIT);

    ASSERT_INT("init: state=GREEN", g.state, AF_GUARD_GREEN);
    ASSERT_NEAR("init: peak=10000", g.peak_equity, 10000.0, 1e-9);

    /* Check at equity=9900: -1 % daily → GREEN */
    s = af_drawdown_guard_check(&g, 9900.0, 10000.0);
    ASSERT_INT("9900 (1% loss) → GREEN", s.state, AF_GUARD_GREEN);
    ASSERT_INT("9900 → trading_allowed=1", s.trading_allowed, 1);

    /* Check at equity=9700: -3 % daily → still GREEN (< 75 % of 5 %) ?
     * warn_thresh = 0.75 * 0.05 = 0.0375 daily loss
     * 9700 daily_pct = (9700-10000)/10000 = -0.03 < 0.0375 → GREEN */
    s = af_drawdown_guard_check(&g, 9700.0, 10000.0);
    ASSERT_INT("9700 (3% loss) → GREEN", s.state, AF_GUARD_GREEN);

    /* Check at equity=9620: -3.8 % → daily_pct=-0.038 ≤ -0.0375 (75% of 5%) → YELLOW */
    s = af_drawdown_guard_check(&g, 9620.0, 10000.0);
    ASSERT_INT("9620 (3.8% loss) → YELLOW", s.state, AF_GUARD_YELLOW);
    ASSERT_INT("9620 → trading_allowed=1", s.trading_allowed, 1);

    /* Check at equity=9400: -6 % daily (> 5 % limit) → RED */
    s = af_drawdown_guard_check(&g, 9400.0, 10000.0);
    ASSERT_INT("9400 (6% loss) → RED", s.state, AF_GUARD_RED);
    ASSERT_INT("9400 → trading_allowed=0", s.trading_allowed, 0);
    ASSERT_NEAR("9400 daily_pnl=-600", s.daily_pnl, -600.0, 1.0);

    /* Daily total-drawdown check: equity=8500 from peak=10000 → 15 % → RED */
    af_drawdown_guard_t g2;
    af_drawdown_guard_init(&g2, 10000.0, 10000.0,
                           AF_DRAWDOWN_DAILY_LIMIT,
                           AF_DRAWDOWN_TOTAL_LIMIT);
    s = af_drawdown_guard_check(&g2, 8500.0, 10000.0);
    ASSERT_INT("8500 (15% total DD) → RED", s.state, AF_GUARD_RED);
    ASSERT_NEAR("8500 total_dd_pct≈0.15", s.total_dd_pct, 0.15, 0.001);

    /* Equity rises above peak → peak updated */
    af_drawdown_guard_t g3;
    af_drawdown_guard_init(&g3, 10000.0, 10000.0,
                           AF_DRAWDOWN_DAILY_LIMIT,
                           AF_DRAWDOWN_TOTAL_LIMIT);
    af_drawdown_guard_check(&g3, 10500.0, 10500.0);
    ASSERT_NEAR("peak updated to 10500", g3.peak_equity, 10500.0, 1e-9);

    /* Recover YELLOW → GREEN when metrics improve */
    af_drawdown_guard_t g4;
    af_drawdown_guard_init(&g4, 10000.0, 10000.0,
                           AF_DRAWDOWN_DAILY_LIMIT,
                           AF_DRAWDOWN_TOTAL_LIMIT);
    af_drawdown_guard_check(&g4, 9620.0, 10000.0); /* → YELLOW (3.8% loss > 3.75% warn) */
    ASSERT_INT("after 9620: YELLOW", g4.state, AF_GUARD_YELLOW);
    s = af_drawdown_guard_check(&g4, 9950.0, 10000.0); /* → back to GREEN */
    ASSERT_INT("after 9950: back to GREEN", s.state, AF_GUARD_GREEN);
}

/* ──────────────────────────────────────────────────────────────────────────
 * main
 * ────────────────────────────────────────────────────────────────────────── */
int main(void)
{
    printf("AlgoForge Round 10 — Risk Unit Tests\n");
    printf("=====================================\n");

    test_kelly();
    test_normalize();
    test_size_position();
    test_drawdown_guard();

    printf("\n=====================================\n");
    printf("Results: %d/%d passed", g_passed, g_tests);
    if (g_failed > 0) {
        printf("  *** %d FAILED ***\n", g_failed);
        return 1;
    }
    printf("  — all OK\n");
    return 0;
}
